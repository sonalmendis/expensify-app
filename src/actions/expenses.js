import uuidv1 from 'uuid/v1';

// ADD EXPENSE ACTION
export const addExpense = ({ description = '', note = '', amount = 0, createdAt = 0 } = {}) => ({
	type: 'ADD_EXPENSE',
	expense: {
		id: uuidv1(),
		description,
		note,
		amount,
		createdAt
	}
});

// REMOVE EXPENSE ACTION
export const removeExpense = ({ id } = {}) => ({
	type: 'REMOVE_EXPENSE',
	id
});

// EDIT EXPENSE ACTION
export const editExpense = ({ id, updates }) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
});
