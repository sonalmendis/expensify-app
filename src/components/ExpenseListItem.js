import React from 'react';

import { Link } from 'react-router-dom';
const ExpenseListItem = ({ description, amount, createdAt, id, dispatch }) => (
	<div>
		<Link to={'/edit/' + id}>
			<h2>{description}</h2>
		</Link>

		<h3>{amount}</h3>
		<h4>{createdAt}</h4>
	</div>
);

export default ExpenseListItem;

/*
The above is the same as:
export default (props) => (
	<div>
		<h2>{props.expenseDesc}</h2>
		<h3>{props.expenseAmount}</h3>
	</div>
);
but just de-structured for easier reading
*/
