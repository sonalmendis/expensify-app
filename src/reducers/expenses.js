const expensesReducerDefaultState = []; // example of setting the default state values as a var beforehand
const expensesReducer = (state = expensesReducerDefaultState, action) => {
	switch (action.type) {
		case 'ADD_EXPENSE':
			return [...state, action.expense];
		// the spread operator [...state] works similar to state.concat(action.expense);
		case 'REMOVE_EXPENSE':
			return state.filter(
				({ id }) => id !== action.id
				// console.log(action.id);
				// console.log(id);
			);
		/* filter above is shorthand for:
                state.filter(({id}) => {
                    return id !== action.id
                })
                ({id}) is just a destructured (expense.id)

        */
		case 'EDIT_EXPENSE':
			return state.map(expense => {
				if (expense.id === action.id) {
					return {
						...expense,
						...action.updates
					};
				} else {
					return expense;
				}
			});
		case 'SET_EXPENSES':
			return action.expenses;
		default:
			return state;
	}
};

export default expensesReducer;
