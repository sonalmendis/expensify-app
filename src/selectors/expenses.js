import moment from 'moment';

// Get visible expense
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
	/*
    The way this function works is by going through the states expenses one by one and then console logging
    them based on the filters from the dispatch. The filters are a way to visibly sort through items BUT
    they don't change the state directly which is why its not used in a reducer.
    */
	return expenses
		.filter(expense => {
			const createdAtMoment = moment(expense.createdAt);
			const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
			// console.log(startDate);
			// console.log(startDateMatch);
			const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
			// console.log(text);

			const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());

			return startDateMatch && endDateMatch && textMatch; // Anything that return false will get filtered out
		})
		.sort((a, b) => {
			//https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
			if (sortBy === 'date') {
				return a.createdAt < b.createdAt ? 1 : -1;
			} else if (sortBy === 'amount') {
				return a.amount < b.amount ? 1 : -1;
			}
		});
};

export default getVisibleExpenses;
