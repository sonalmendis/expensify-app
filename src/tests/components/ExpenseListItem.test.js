import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('should render expense list item with single expense', () => {
	console.log(expenses[0]);

	const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
	expect(wrapper).toMatchSnapshot();
});
