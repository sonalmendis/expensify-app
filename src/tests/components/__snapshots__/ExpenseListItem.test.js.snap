// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list item with single expense 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h2>
      Gum
    </h2>
  </Link>
  <h3>
    500
  </h3>
  <h4>
    0
  </h4>
</div>
`;
